File: src/Rectangle.cpp
#include "Rectangle.h"

Rectangle::Rectangle() : x_(0), y_(0), w_(0), h_(0) {}
Rectangle::Rectangle(double x, double y, double w, double h)
    : x_(x), y_(y), w_(w), h_(h) {}

bool Rectangle::contains(Point const &p) const noexcept {
  return (p.x > (x_ - w_)) && (p.x < (x_ + w_)) && (p.y > (y_ - h_)) &&
         (p.y < (y_ + h_));
}

bool Rectangle::intersects(Rectangle const &r) const noexcept {
  return !(r.x() - r.w() > x_ + w_ || r.x() + r.w() < x_ - w_ ||
           r.y() - r.h() > y_ + h_ || r.y() + r.h() < y_ - h_);
}

double Rectangle::x() const noexcept { return x_; }
double Rectangle::y() const noexcept { return y_; }
double Rectangle::w() const noexcept { return w_; }
double Rectangle::h() const noexcept { return h_; }

File: python/Rectangle.py
class Rectangle:
    def __init__(self, center, width, height):
        self.center = center
        self.width = width
        self.height = height

    def x(self):
        return self.center.x

    def y(self):
        return self.center.y

    def contains(self, other):
        return (other.x > (self.x() - self.width))\
            and (other.x < (self.x() + self.width))\
            and (other.y > (self.y() - self.height))\
            and (other.y < (self.y() + self.height))

    def intersects(self, other):
        return not (other.x() - other.width > self.x() + self.width
                    or other.x() + other.width < self.x() - self.width
                    or other.y() - other.height > self.y() + self.height
                    or other.y() + other.height < self.y() - self.height)
